include/ignition/gui/PlottingInterface.hh

```diff
@@ -0,0 +1,203 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+#ifndef IGNITION_GUI_PLOTTINGINTERFACE_HH_
+#define IGNITION_GUI_PLOTTINGINTERFACE_HH_
+
+#include <QObject>
+#include <QTimer>
+#include <QString>
+#include <QVariant>
+#include <bits/stdc++.h>
+#include <google/protobuf/message.h>
+#include <google/protobuf/descriptor.h>
+#include <ignition/transport/Node.hh>
+#include <ignition/transport/MessageInfo.hh>
+#include <ignition/transport/Publisher.hh>
+#include <map>
+#include <set>
+#include <string>
+#include <memory>
+
+namespace ignition
+{
+namespace gui
+{
+class Field
+{
+  /// \brief value of that field
+  private: double value;
+
+  /// \brief Registered Charts to that field
+  private: std::set<int> charts;
+
+  /// \brief Constructor
+  public: Field();
+
+  /// \param[in] _value the set value
+  public: void SetValue(double _value);
+
+  /// \return value of the field
+  public: double Value();
+
+  /// \brief Register a chart that plot that field
+  public: void AddChart(int _chart);
+
+  /// \brief UnRegister a chart from plotting that field
+  /// \return the count of registered charts
+  public: void RemoveChart(int _chart);
+
+  /// \brief size of registered charts
+  /// \return charts size
+  public: int ChartCount();
+
+  public: std::set<int> &Charts();
+};
+
+class Topic
+{
+  /// \brief topic name
+  private: std::string name;
+
+  /// \brief Plotting fields to update its values
+  private: std::map<std::string, Field*> fields;
+
+  /// \brief Constructor
+  public: explicit Topic(std::string _name);
+
+  /// \brief get topic name
+  public: std::string Name();
+
+  /// \brief Register a chart to a field
+  /// \param[in] _fieldPath model path to the field as an ID
+  public: void Register(std::string _fieldPath, int _chart);
+
+  /// \brief remove field from the topic
+  public: void UnRegister(std::string _fieldPath, int _chart);
+
+  /// \brief size of registered fields
+  /// \return fields size
+  public: int FieldCount();
+
+  /// \brief get the registered topics
+  /// \return topics list
+  public: std::map<std::string, Field*> &Fields();
+  /// \brief callback to subscribe to that topic
+  /// \param[in] _msg the published msg from the topic
+  public: void Callback(const google::protobuf::Message &_msg);
+
+  /// \brief check the plotable types and get data from reflection
+  private: double PlotData(const google::protobuf::Message &_msg,
+                           const google::protobuf::FieldDescriptor *field);
+};
+
+class TransportPrivate;
+
+/// \brief handle Transport Topics Subscribing for one object(Chart)
+class Transport
```

Comment:
Can the `Field`, `Topic` and `Transport` classes be moved to another `.hh` file under the `src` folder? The least we have in this header the better, because everything under `include` in installed and can't be changed after it is released.

Reply:
ok but how can i include them after moving them to a .hh file in the /src folder ?

Reply:
For example:

* `include/ignition/gui/PlottingInterface.hh`
    * Declares `PlottingInterface`
    * Doesn't need `Field`, `Topic` or `Transport`
* `src/PlottingTransport.hh`
   * Declares `Field`, `Topic` and `Transport` (just like you have here, but on a separate file)
* `src/PlottingTransport.cc`
    * Implements `Field`, `Topic` and `Transport`
    * Includes `#include "PlottingTransport.hh"`
* `src/PlottingInterface.cc`
    * Includes `#include "PlottingTransport.hh"`
    * Includes `#include "ignition/gui/PlottingInterface.hh"`
    * Implements `PlottingInterface`
    * Uses `Field`, `Topic` or `Transport`